include/storage/fbw.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for flash buffered writes
+ */
+
+#ifndef ZEPHYR_INCLUDE_STORAGE_FBW_H_
+#define ZEPHYR_INCLUDE_STORAGE_FBW_H_
+
+/**
+ * @brief Abstraction over flash buffered writes.
+ *
+ * @defgroup fbw flash buffered write Interface
+ * @{
+ */
+
+#include <stdbool.h>
+#include <storage/flash_map.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct fbw_ctx {
+	u8_t *buf;
+	size_t buf_len;
+	const struct flash_area *flash_area;
+	size_t bytes_written;
+	u16_t buf_bytes;
+#ifdef CONFIG_IMG_ERASE_PROGRESSIVELY
+	off_t off_last;
+#endif
+};
+
+/**
+ * @brief Initialize context needed for buffered write to flash.
```

Comment:
When you are ready to take this out of draft status, I think it'd be good to mark up these API functions according to the new [terminology](https://docs.zephyrproject.org/latest/reference/index.html#terminology) introduced by @pabigot 

For example, I guess this one is pre-kernel-ok and not reschedule.

Reply:
I will make sure to mark them up (Y)

Reply:
I'm not able to find any example usage of those tags, could you point me to a file where this is already done?

Reply:
You're a guinea pig :)

Reply:
someone have to be the firs one

Reply:
I tried adding some tags. Really unsure about the syntax that should be used (is there any?).

Also, I'm not sure if flash operations are available pre-kernel? If not then I guess all my proposed tags are wrong, so please check those xD

Reply:
> When you are ready to take this out of draft status, I think it'd be good to mark up these API functions according to the new [terminology](https://docs.zephyrproject.org/latest/reference/index.html#terminology) introduced by @pabigot

Also CC @carlescufi -- can we add this to the API agenda for next week as a test run?

Reply:
API meeting:

Suggestion from @nashif:
- Use a doxygen alias to map something like `api-tags` to a regular Doxygen `note` that can be parsed by breathe.

@mbolivar-nordic  @hakonfam: let's  use `note` for now, and we will revisit this anyway once the above is resolved

Reply:
Added @note, please check if it looks correct

Reply:
I would vote for `@note pre-kernel-ok` or so, but will leave this up to @carlescufi and @pabigot

Reply:
I have no strong opinions about the style associated with documenting API behavior, except that I would avoid making things too strained by attempting to prepare for some future machine processing that is unlikely to exist anytime soon.  "This function is isr-ok and not sleep." covers the core need, with or without a preceding `@note`.  (Said future tooling should convert `sleep` to [sleep](https://docs.zephyrproject.org/latest/reference/index.html#sleep) to make clear it's being used as an adjective/attribute, not a verb.)

For the content: [pre-kernel-ok](https://docs.zephyrproject.org/latest/reference/index.html#pre-kernel-ok) is one of the less useful tags.  It's really only important to know if the API is something that could reasonably be expected to be invoked before startup switches from the interrupt stack to a thread stack.  In that case it almost certainly would be paired with [isr-ok](https://docs.zephyrproject.org/latest/reference/index.html#isr-ok).

Developers also need to be aware that both `pre-kernel-ok` and `isr-ok` in isolation can be satisfied by having the function return an error if called from a context where they aren't allowed to sleep.   Both should be paired with `not sleep` to suggest that the function actually does its job when invoked from a non-thread context.

AFAICT all flash drivers initialize post-kernel, so if this function requires a flash device it's pretty unlikely that `pre-kernel-ok` is relevant.

Also don't forget to document expectations for the `cb` parameter: does this API expect it to be invoked from an interrupt handler?  That might be best documented at the typedef for `fbw_callback_t`, which currently says nothing about what it is or how it's used.

Reply:
I removed the `pre-kernel-ok` for all funtions except `fbw_bytes_written` which has both `pre-kernel-ok` and `isr-ok`. 

As for `fbw_callback_t` I added docs to that typedef now, but it is not clear to me how I should explain where the "API expect it to be invoked from an interrupt handler", so if we want that you would have to help me with the phrasing.

Reply:
From what I can tell `fbw_oper` inherits the API behavior of `flash_oper` given whatever flash device is provided in the context.  Some will be **sleep**, few if any will be **isr-ok**.

Reply:
OK, then I suggest that we remove the API-tags altogether, as it is difficult to give any information of value.